### ISSUE TITLE
L2 Basic Iterative Attack returns infeasible perturbation

### ISSUE THREAD

**Issue by shinington** [bethgelab/foolbox#388]
Hi, 

I'm trying to use L2BasicIterativeAttack to produce adversarial examples but get a perturbation the L2 norm of which is much larger than the epsilon I set. I made a minimal modification on the Pytorch example in the tutorial and the only difference is the attack model. Could you have a look? Thanks! My test code is below:

```
import foolbox
import torch
import torchvision.models as models
import numpy as np

from numpy.linalg import norm

# instantiate the model
resnet18 = models.resnet18(pretrained=True).eval()
if torch.cuda.is_available():
    resnet18 = resnet18.cuda()
mean = np.array([0.485, 0.456, 0.406]).reshape((3, 1, 1))
std = np.array([0.229, 0.224, 0.225]).reshape((3, 1, 1))
fmodel = foolbox.models.PyTorchModel(
    resnet18, bounds=(0, 1), num_classes=1000, preprocessing=(mean, std))

# get source image and label
image, label = foolbox.utils.imagenet_example(data_format='channels_first')
image = image / 255.  # because our model expects values in [0, 1]

print('label', label)
print('predicted class', np.argmax(fmodel.forward_one(image)))

# apply attack on source image
metric = foolbox.distances.MSE
attack = foolbox.attacks.L2BasicIterativeAttack(fmodel, distance=metric)

eps = 0.03
step_size = eps/4
steps = 20

adversarial = attack(image, label, binary_search=False, epsilon=eps, 
                             stepsize=step_size, iterations=steps, random_start=True)

print('adversarial class', np.argmax(fmodel.forward_one(adversarial)))

perturbation = adversarial - image
print('Max perturbation: {}'.format(np.amax(perturbation)))
print('Min perturbation: {}'.format(np.amin(perturbation)))
print('L2 norm of perturbation: {}'.format(norm(perturbation.flatten(), 2)))
```
And the result is 

```
label 282
predicted class 282
adversarial class 728
Max perturbation: 0.08541756868362427
Min perturbation: -0.09411180019378662
L2 norm of perturbation: 7.301252841949463
```
The L2 norm of the perturbation (~7.30) is much greater than epsilon=0.03. BTW, I'm using the latest version of foolbox.

**Comment by persistz** [comment]
I think you need to transfer between epsilon=0.03 and L2 norm, 
epsilon=0.03 means Linf norm.
The calculation formula is different, or they are totally different things.
If you assume every pixels change 0.3, you can get an L2 norm= (299 * 299 * 3 * 0.3 * 0.3)^(1/2), which about 150.

**Comment by shinington** [comment]
> I think you need to transfer between epsilon=0.03 and L2 norm,
> epsilon=0.03 means Linf norm.
> The calculation formula is different, or they are totally different things.
> If you assume every pixels change 0.3, you can get the L2 norm value, which = (299 * 299 * 3 * 0.3 * 0.3)^(1/2), about 150.
> 7.30 is a small value for L2 norm on ImageNet dataset.

Thanks for the prompt reply. Isn't epsilon an argument of the L2BasicIterativeAttack? Which means it should be the max L2 perturbation?

**Comment by jonasrauber** [comment]
Quote from the [docs](https://foolbox.readthedocs.io/en/latest/modules/v1_attacks/gradient.html#foolbox.attacks.L2BasicIterativeAttack): 
> epsilon:float
> Limit on the perturbation size; if binary_search is True, this value is only for initialization and automatically adapted.

The idea in Foolbox is to find the smallest perturbation such that the perturbed image is an adversarial. One can then compare the size of this perturbation to different thresholds. Looking at a single threshold is usually not a good idea and often leads to misleading evaluations of robustness.

**Comment by shinington** [comment]
> Quote from the [docs](https://foolbox.readthedocs.io/en/latest/modules/v1_attacks/gradient.html#foolbox.attacks.L2BasicIterativeAttack):
> 
> > epsilon:float
> > Limit on the perturbation size; if binary_search is True, this value is only for initialization and automatically adapted.
> 
> The idea in Foolbox is to find the smallest perturbation such that the perturbed image is an adversarial. One can then compare the size of this perturbation to different thresholds. Looking at a single threshold is usually not a good idea and often leads to misleading evaluations of robustness.

Thanks for your reply. One more question: please note that in my test code above binary_search=False so there is no binary search for the smallest perturbation and here epsilon is the "limit on the perturbation size". If my understanding is correct, this means that the L2 norm of the resulting perturbation should be less than epsilon right? But running my test code abobe showed the opposite.

If my understanding is incorrect, do you have any suggestions if I want to find a perturbation whose L2 norm is limited by a value?

**Comment by jonasrauber** [comment]
Sorry, I didn't notice that you actually use `binary_search=False`.
I think the `L2BasicIterativeAttack` currently uses a renormalized `norm` and compares `epsilon` against that one. This renormalization takes both the bounds and the image size into account (and is supposed to make the hyperparameters more independent of these factors).
In this case, however, I admit that it makes them less intuitive.

In this case, it effectively compares `epsilon` against `sqrt(mean(square(perturbation)))`. This means you should divide `epsilon` and `stepsize` by `sqrt(height * width)`, which is 224 in your case.

```python3
eps = 0.03 / 224
step_size = eps/4 / 224
```

This should do the trick. Let me know if it helped and please double check the numbers.

**Comment by shinington** [comment]
> Sorry, I didn't notice that you actually use `binary_search=False`.
> I think the `L2BasicIterativeAttack` currently uses a renormalized `norm` and compares `epsilon` against that one. This renormalization takes both the bounds and the image size into account (and is supposed to make the hyperparameters more independent of these factors).
> In this case, however, I admit that it makes them less intuitive.
> 
> The attack effectively compares `epsilon` against `sqrt(mean(square(perturbation)))`. This means you should divide `epsilon` and the `stepsize` by `sqrt(height * width)`, which is 224 in your case (and multiply by `max_ - min_`, i.e. the bounds, but they are `1` in your case).
> 
> ```python
> eps = 0.03 / 224
> step_size = eps/4 / 224
> ```
> 
> This should do the trick. Let me know if it helped and please double check the numbers.

Thanks for the reply. I did the modification as suggested but the result is kind of weird: I increased the value of ```epsilon``` to be 5/224 to make it more reasonable (here 0.03/224 is too small to generate an *evasive* adversarial example), with ```step_size = epsilon/4``` and ```steps = 20```. The L2 norm of the resulting perturbation should be less than 5, but it was actually larger than that. The code is below:

```
import foolbox
import torch
import torchvision.models as models
import numpy as np

from numpy.linalg import norm

# instantiate the model
resnet18 = models.resnet18(pretrained=True).eval()
if torch.cuda.is_available():
    resnet18 = resnet18.cuda()
mean = np.array([0.485, 0.456, 0.406]).reshape((3, 1, 1))
std = np.array([0.229, 0.224, 0.225]).reshape((3, 1, 1))
fmodel = foolbox.models.PyTorchModel(
    resnet18, bounds=(0, 1), num_classes=1000, preprocessing=(mean, std))

# get source image and label
image, label = foolbox.utils.imagenet_example(data_format='channels_first')
image = image / 255.  # because our model expects values in [0, 1]

print('label', label)
print('predicted class', np.argmax(fmodel.forward_one(image)))

# apply attack on source image
metric = foolbox.distances.MSE
attack = foolbox.v1.attacks.L2BasicIterativeAttack(fmodel, distance=metric)

eps = 5/224
step_size = eps/4
steps = 20

adversarial = attack(image, label, binary_search=False, epsilon=eps, 
                             stepsize=step_size, iterations=steps, random_start=True)

print('adversarial class', np.argmax(fmodel.forward_one(adversarial)))

perturbation = adversarial - image
print('L2 norm of perturbation: {}'.format(norm(perturbation.flatten(), 2)))
```
And the result is

```
label 282
predicted class 282
adversarial class 333
L2 norm of perturbation: 5.440914154052734
```
The L2 norm of perturbation is ~5.44 > 5.

**Comment by jonasrauber** [comment]
Sorry, I forgot about the three color channels. The mean is over `224 * 224 * 3` elements. So `eps` should be divided by `np.sqrt(224 * 224 * 3)`:

`eps = 5 / np.sqrt(224 * 224 * 3)`

Then the L2 norm will be smaller than `5`, in my case `3.13`.

**Comment by shinington** [comment]
> Sorry, I forgot about the three color channels. The mean is over `224 * 224 * 3` elements. So `eps` should be divided by `np.sqrt(224 * 224 * 3)`:
> 
> `eps = 5 / np.sqrt(224 * 224 * 3)`
> 
> Then the L2 norm will be smaller than `5`, in my case `3.13`.

This helps! After using the code above now the perturbation becomes feasible. I've used a batch of 100 samples to verify this.

It would be highly appreciated if there could be more annotations about the use of epsilon in the document when L2 attacks are applied. In fact, ```epsilon``` here is only an interpretation of the equivalent Linf perturbation, but not 'limit on the perturbation size' as described in the document.

Here I have one more question about the random start of L2 Basic Iterative Attack(and I hope this will not bother you guys too much): according to Line 171 to 177 and Line 276 to 285 in ```foolbox/v1/attacks/iterative_projected_gradient.py```, when L2 norm and random start is applied, the initial noise is not uniformly sampled from the L2 ball, instead, it is uniformly sampled from the largest hypercube inside the ball. Considering that in a very high dimensional space the most volume of an L2 ball concentrates near its surface while the most volume of a hypercube concentrates near its corners, does the use of such sampling result in a weaker attack since most of the volume is not sampled? Please feel free to let me know if my understanding is incorrect. Thanks in advance!

Moreover, Line 172 and 173 of the same code tells that 'using uniform noise even if the perturbation clipping uses a different norm because cleverhans does it the same way', but actually this is not cleverhans does in its lates version. According to Line 496 to 503 of ```cleverhans/utils_tf.py```, they use a sampling method which is roughly equivalent to sampling from the surface of the ball.

**Comment by jonasrauber** [comment]
> In fact, epsilon here is only an interpretation of the equivalent Linf perturbation, but not 'limit on the perturbation size' as described in the document.

`epsilon` in this case is a limit on the (l2) perturbation size, the unit of `epsilon` is just not the raw `l2` norm but a rescaled form (rescaled by `sqrt` of the image size (`height * width * channels`) and `max - min` where `min` and `max` are the bounds of the pixel values)

But you are right, it would certainly be good to better document this (or to change the meaning of `epsilon` to simply be the `l2` norm).

As the comment in the code shows, this was done to follow what cleverhans did. If that changed (or there are other good reasons to change this), I am happy to accept a pull-request that changes this in our code as well. @shinington Could you create one with the necessary changes?

**Comment by shinington** [comment]
> > In fact, epsilon here is only an interpretation of the equivalent Linf perturbation, but not 'limit on the perturbation size' as described in the document.
> 
> `epsilon` in this case is a limit on the (l2) perturbation size, the unit of `epsilon` is just not the raw `l2` norm but a rescaled form (rescaled by `sqrt` of the image size (`height * width * channels`) and `max - min` where `min` and `max` are the bounds of the pixel values)
> 
> But you are right, it would certainly be good to better document this (or to change the meaning of `epsilon` to simply be the `l2` norm).
> 
> As the comment in the code shows, this was done to follow what cleverhans did. If that changed (or there are other good reasons to change this), I am happy to accept a pull-request that changes this in our code as well. @shinington Could you create one with the necessary changes?

Sure. I would like to. Thanks a lot for the reply.
